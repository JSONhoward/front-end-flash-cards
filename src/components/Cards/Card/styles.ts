import styled, {keyframes, css} from 'styled-components';

const animation = keyframes`
from {
    transform: translate(-50%, -50%);
}
to {
    transform: translate(-400%, -50%) rotate(-90deg);
}
`

export const Loading = styled('div')`
position: absolute;
top: 50%;
left: 50%;
transform: translate(-50%, -50%);
display: flex;
flex-direction: column;
align-items: center;
justify-content: space-around;
height: 30vw;
width: 50vw;
min-height: 168px;
min-width: 280px;
padding: 4rem;
background-color: whitesmoke;
box-shadow: 2px 2px 10px rgba(0,0,0,.5);
font-family: 'Permanent Marker', cursive;
font-size: 1.5rem;

p {
    text-align: center;
}
`

export const AnswerText = styled('p')<{show: boolean}>`
filter: ${({show}) => !show ? `blur(5px)` : 'none'};
font-size: 2.5vw;
font-style: italic;
font-family: 'Trebuchet MS', 'Lucida Sans Unicode', 'Lucida Grande', 'Lucida Sans', Arial, sans-serif;

@media screen and (min-width: 600px) {
 font-size: 1.25rem;
}
`

export const FlashCard = styled('div')<{offset: number, done: boolean}>`
position: absolute;
top: 50%;
left: 50%;
transform: ${({offset}) => `translate(${-50 + offset}%, ${-50 + offset}%)`};
display: flex;
flex-direction: column;
align-items: center;
justify-content: space-around;
height: 30vw;
width: 50vw;
min-height: 168px;
min-width: 280px;
padding: .5rem;
background-color: whitesmoke;
box-shadow: 2px 2px 10px rgba(0,0,0,.5);
animation: ${({done}) => done ? css`${animation} 1s linear` : 'none'};
font-family: 'Permanent Marker', cursive;
font-size: 1.5rem;

p {
    text-align: center;
    font-size: 3vw;
    white-space: pre-wrap;
}

code {
    background-color: black;
    max-height: 50%;
    width: fit-content;
    min-width: 75%;
    color: white;
    font-family: monospace;
    font-size: .5rem;
    white-space: pre-wrap;
    overflow: auto;
}

@media screen and (min-width: 600px) {
    p {
        font-size: 1.5rem;
    }
    code {
        font-size: 1.25rem;
    }
}
`

