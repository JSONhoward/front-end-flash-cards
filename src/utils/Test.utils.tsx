import * as React from 'react'
import { RecoilRoot } from 'recoil'

export const RecoilWrapper: React.FC = ({children}) => {
    return (
        <RecoilRoot>
            {children}
        </RecoilRoot>
    )
}