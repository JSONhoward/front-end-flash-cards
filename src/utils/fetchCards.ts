export const fetchCards = async (): Promise<FirebaseDocArray> => {
    const response = await fetch(process.env.REACT_APP_GRAPHQL_URL as string, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'X-REQUEST-TYPE': 'GraphQL'
        },
        body: JSON.stringify({
            query: fetchAllCardsQuery
        })
    })
    const data = await response.json()
    const {data: {cards}} = data

    for (let i = cards.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [cards[i], cards[j]] = [cards[j], cards[i]]
    }

    return cards
}

const fetchAllCardsQuery = `
query {
    cards {
        id
        Category
        q
        code
        a
    }
}
`