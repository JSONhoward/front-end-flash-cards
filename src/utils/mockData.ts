export const mockQuestionArray: FirebaseDocArray = [
    {
        id: '1',
        Category: 'General',
        q: 'When building a new web site or maintaining one, can you explain some techniques you have used to increase performance?',
        a: 'some answer'
    },
    {
        id: '2',
        Category: 'General',
        code: 'some code',
        q: 'Test question 2',
        a: 'some answer'
    },
    {
        id: '3',
        Category: 'HTML',
        code: 'some code',
        q: 'Test question 3',
        a: 'some answer'
    },
    {
        id: '4',
        Category: 'Javascript',
        q: 'Test question 4',
        a: 'some answer'
    },
    {
        id: '5',
        Category: 'CSS',
        q: 'Test question 5',
        a: 'some answer'
    },
    {
        id: '6',
        Category: 'Networking',
        q: 'Test question 6',
        a: 'some answer'
    },
    {
        id: '7',
        Category: 'Testing',
        q: 'Test question 7',
        a: 'some answer'
    },
    {
        id: '8',
        Category: 'Coding',
        q: 'Test question 8',
        a: 'some answer'
    },
    {
        id: '9',
        Category: 'General',
        q: 'Test question 9',
        a: 'some answer'
    },
    {
        id: '10',
        Category: 'General',
        q: 'Test question 10',
        a: 'some answer'
    }
]