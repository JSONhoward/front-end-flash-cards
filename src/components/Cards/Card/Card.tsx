import React from 'react'
import { FaEyeSlash, FaEye, FaArrowRight } from 'react-icons/fa'

import { AnswerText, FlashCard, RemoveCardButton, ShowAnswerButton, Loading } from './styles';

type Props = {
    question: string,
    answer: string,
    remove: boolean,
    handleRemove: () => void,
    handleShow: () => void
    rotation: number,
    show: boolean,
    loading: boolean
}

const Card: React.FC<Props> = ({ question, answer, rotation, handleRemove, handleShow, remove, show, loading }) => {
    return (
        <>
            <FlashCard done={remove} rotation={rotation}>
                {
                    loading ? <Loading rotation={rotation}>Loading...</Loading> : (
                        <>
                            <p>{question}</p>
                            <AnswerText show={show}>{answer}</AnswerText>
                            <RemoveCardButton onClick={handleRemove}>
                                <FaArrowRight size={'2rem'} />
                            </RemoveCardButton>
                            <ShowAnswerButton show={show} onClick={handleShow}>
                                {
                                    show ? <FaEye size={'2rem'} /> : <FaEyeSlash size={'2rem'} />
                                }
                            </ShowAnswerButton>
                        </>
                    )
                }
            </FlashCard>
        </>
    )
}

export default Card
