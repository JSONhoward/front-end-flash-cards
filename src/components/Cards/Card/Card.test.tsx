import React from 'react'
import { render, screen, cleanup } from '@testing-library/react';
import { create } from 'react-test-renderer'
import 'jest-styled-components'

import { RecoilWrapper } from '../../../utils/Test.utils';
import Card from './Card'

const props = {
    question: 'test question',
    answer: 'test answer',
    remove: false,
    offset: 0,
    show: false,
    loading: false,
    code: 'test code'
}

afterEach(cleanup)

describe('Card Tests', () => {
    it('Matches Snapshot', () => {
        const component = create(<RecoilWrapper><Card {...props} /></RecoilWrapper>)

        expect(component.toJSON()).toMatchSnapshot()
    })

    it('Renders card correctly', () => {
        render(<RecoilWrapper><Card {...props} /></RecoilWrapper>)

        expect(screen.getByText(props.question).textContent).toBe(props.question)
        expect(screen.getByText(props.code).textContent).toBe(props.code)
        expect(screen.getByText(props.answer).textContent).toBe(props.answer)
    })

    it('Hides(blurs) answer', () => {
        render(<RecoilWrapper><Card {...props} /></RecoilWrapper>)

        expect(screen.getByText(props.answer)).toHaveStyleRule('filter', 'blur(5px)')
    })
})