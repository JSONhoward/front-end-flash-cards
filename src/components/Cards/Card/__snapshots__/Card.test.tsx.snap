// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Tests Matches Snapshot 1`] = `
.c1 {
  -webkit-filter: blur(5px);
  filter: blur(5px);
  font-size: 2.5vw;
  font-style: italic;
  font-family: 'Trebuchet MS','Lucida Sans Unicode','Lucida Grande','Lucida Sans',Arial,sans-serif;
  max-height: 50%;
  overflow-y: auto;
  overflow-x: hidden;
}

.c0 {
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  height: 30vw;
  width: 50vw;
  min-height: 168px;
  min-width: 280px;
  padding: .5rem;
  background-color: whitesmoke;
  box-shadow: 2px 2px 10px rgba(0,0,0,.5);
  -webkit-animation: none;
  animation: none;
  font-family: 'Permanent Marker',cursive;
  font-size: 1.5rem;
}

.c0 p {
  text-align: center;
  font-size: 3vw;
  white-space: pre-wrap;
}

.c0 code {
  background-color: black;
  max-height: 50%;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  min-width: 75%;
  color: white;
  font-family: monospace;
  font-size: .5rem;
  white-space: pre-wrap;
  overflow: auto;
}

@media screen and (min-width:600px) {
  .c1 {
    font-size: 1.25rem;
  }
}

@media screen and (min-width:600px) {
  .c0 p {
    font-size: 1.5rem;
  }

  .c0 code {
    font-size: 1.25rem;
  }
}

<div
  className="c0"
  offset={0}
>
  <p>
    test question
  </p>
  <code>
    test code
  </code>
  <p
    className="c1"
  >
    test answer
  </p>
</div>
`;
